eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/RelayerSpec.scala

```diff
@@ -0,0 +1,196 @@
+import fr.acinq.eclair.crypto.Sphinx
+import fr.acinq.eclair.payment.IncomingPacket.FinalPacket
+import fr.acinq.eclair.payment.OutgoingPacket.{Upstream, buildCommand}
+import fr.acinq.eclair.payment.PaymentPacketSpec.{c, channelUpdate_bc, finalAmount, finalExpiry, hops, makeCommitments, paymentHash}
+import fr.acinq.eclair.payment.relay.Relayer.{ChildActors, GetChildActors, RelayForward}
+import fr.acinq.eclair.payment.{OutgoingPacket, PaymentPacketSpec}
+import fr.acinq.eclair.router.Router.{ChannelHop, NodeHop}
+import fr.acinq.eclair.wire.Onion.FinalLegacyPayload
+import fr.acinq.eclair.wire.{FailureMessageCodecs, Onion, RequiredNodeFeatureMissing, UpdateAddHtlc, UpdateFailHtlc, UpdateFulfillHtlc}
+import fr.acinq.eclair.{NodeParams, TestConstants, randomBytes32, _}
+import org.scalatest.funsuite.FixtureAnyFunSuiteLike
+import org.scalatest.{Outcome, Tag}
+
+import scala.concurrent.duration.DurationInt
+
+class RelayerSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("application")) with FixtureAnyFunSuiteLike {
+
+  case class FixtureParam(nodeParams: NodeParams, relayer: akka.actor.ActorRef, router: TestProbe[Any], register: TestProbe[Any], childActors: ChildActors, paymentHandler: TestProbe[Any])
+
+  override def withFixture(test: OneArgTest): Outcome = {
+    eventually {
+      // we are node B in the route A -> B -> C -> ....
+      val disableTrampoline = test.tags.contains("trampoline-disabled")
+      val nodeParams = TestConstants.Bob.nodeParams.copy(enableTrampolinePayment = !disableTrampoline)
+      val router = TestProbe[Any]("router")
+      val register = TestProbe[Any]("register")
+      val paymentHandler = TestProbe[Any]("payment-handler")
+      val probe = TestProbe[Any]()
+      // we can't spawn top-level actors with akka typed
+      testKit.spawn(Behaviors.setup[Any] { context =>
+        val relayer = context.toClassic.actorOf(Relayer.props(nodeParams, router.ref.toClassic, register.ref.toClassic, paymentHandler.ref.toClassic))
+        probe.ref ! relayer
+        Behaviors.empty[Any]
+      })
+      val relayer = probe.expectMessageType[akka.actor.ActorRef]
+      relayer ! GetChildActors(probe.ref.toClassic)
+      val childActors = probe.expectMessageType[ChildActors]
+      withFixture(test.toNoArgTest(FixtureParam(nodeParams, relayer, router, register, childActors, paymentHandler)))
+    }
+  }
+
+  val channelId_ab = randomBytes32
+  val channelId_bc = randomBytes32
+
+  test("relay an htlc-add") { f =>
+    import f._
+
+    // we use this to build a valid onion
+    val (cmd, _) = buildCommand(ActorRef.noSender, Upstream.Local(UUID.randomUUID()), paymentHash, hops, FinalLegacyPayload(finalAmount, finalExpiry))
+    // and then manually build an htlc
+    val add_ab = UpdateAddHtlc(channelId = randomBytes32, id = 123456, cmd.amount, cmd.paymentHash, cmd.cltvExpiry, cmd.onion)
+
+    relayer ! RelayForward(add_ab)
+
+    register.expectMessageType[Register.Forward[CMD_FAIL_HTLC]]
```

Comment:
Can't we emit a `LocalChannelUpdate` like we did before to have this be a `CMD_ADD_HTLC`? Or maybe since this is already tested in `ChannelRelayerSpec`, just assert that the `replyTo` of the `Register.Forward` is a `ChannelRelay` actor?

Reply:
Fixed in 626b4027f.

Reply:
The `expectMessageType` is inside a `println`, was it for your local debugging or is it expected?

Reply:
Oops, fixed in 0b2e5bd7b.